util/EIP-712/src/eip712.rs

```diff
@@ -0,0 +1,135 @@
+use serde_json::{Value};
+use serde::de::{self, Deserialize, Deserializer, Visitor, MapAccess};
+use std::fmt;
+use std::collections::HashMap;
+
+pub type MessageTypes = HashMap<String, Vec<FieldType>>;
+
+#[derive(Debug, Clone)]
+pub struct EIP712 {
+	pub types: MessageTypes,
+	pub primary_type: String,
+	pub message: Value,
+	pub domain: Value,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct FieldType {
+	pub name: String,
```

Comment:
We should make sure that:
1. Every `name` does not have any spaces or commas (otherwise you could mess up the encoding and do: `Address(string na, uint256 me)` and instead of encoding one field you encode two (i.e. the encoding becomes ambigous)
2. The type is one from the list of known/supported types - might be good to actually have an enum for that if possible.

Reply:
@ordian thanks a lot for that crate :handshake:

Reply:
@seunlanlege I think you still need to call `.validate()` to perform validation - I haven't seen it anywhere in the code. If I'm mistaken please add a test case that fails to deserialize invalid values.